src/lib/es5.d.ts

```diff
@@ -584,8 +584,8 @@ interface NumberConstructor {
 /** An object that represents a number of any kind. All JavaScript numbers are 64-bit floating-point numbers. */
 declare var Number: NumberConstructor;
 
-interface TemplateStringsArray extends ReadonlyArray<string> {
-    readonly raw: readonly string[];
+interface TemplateStringsArray extends ReadonlyArray<string | undefined> {

```

Comment:
I think this part is too big of a change; I would keep it as `string` and let people opt into this behavior with `noUncheckedIndexedAccess`.

Reply:
According to my understanding, `noUncheckedIndexedAccess` is used to let the user to ensure the index is not out of length, not applicable for the use case that the item value could really be `undefined`.

This can make the behavior in TypeScript more consistent:

```ts
const array = [ '' ] as string[];
array[array.length - 1].length; // this warning depends on whether noUncheckedIndexedAccess is true

const tuple = [ '' ] as [ string, string | undefined ];
tuple[0].length; // this should not warn even if noUncheckedIndexedAccess is true
tuple[1].length; // this should warn even if noUncheckedIndexedAccess is false

// so TemplateStringsArray is best to be consistent with this behavior
```

Reply:
According to my understanding, `noUncheckedIndexedAccess` is used to let the user to ensure the index is not out of length, not applicable for the use case that the item value could really be `undefined`.

This can make the behavior in TypeScript more consistent:

```ts
const array :string[] = [ '' ];
array[0].length; // this warning depends on whether noUncheckedIndexedAccess is true
```

```ts
const array :( string | undefined )[] = [ '', , '' ];
tuple[1].length; // this warn even if noUncheckedIndexedAccess is false
```

```ts
const tuple :[ string ]= [ '' ];
tuple[0].length; // this will not warn even if noUncheckedIndexedAccess is true
```

So TemplateStringsArray and raw are best to be consistent with these behaviors.

---

a use case for the tuple change:

```ts
function fn ({ raw } :TemplateStringsArray, ...values :string[]) {
    let ret = raw[0];// no need to write `raw[0]!` here any more. more auto, more safe.
    for ( let index = 0; index<values.length; ++index ) {
        ret += values[index]!;
        ret += raw[++index]!;
    }
    return ret;
}
```